Feature/game logic

timjacksonm opened the pull request:
Summary of changes.
-created new level data file that contains 10 levels. First 3 levels are pre-determined who will be the winner. level 4+ is random.
-moved all dino svgs into file that contains level data.
-updated legend component to generate based on level data received from props.
-dinos mapped to battlefield updated with level data. I did move the styles into a new function dinoPerLvlRange. This function duplicates a dino if the range selected is more than #1. Made sense adding styles here so even duplicate dinos are randomly placed.

timjacksonm committed:
Added blue team dino svgs to legend.

timjacksonm committed:
created level 1 object data, added a new dino type with value of three for each team, adjusted legend to receive level data and display appropriate dinos and their values.

timjacksonm committed:
level updates
-Moved level data object into its own file.
-Added levelData state which determines the level that should be played based on url.
-Adjusted redDinos/blueDinos variables to include levelData. dinoType is the svg image, qty for quantity of that dinoType to put on the field. Removed styling and made a top level variable placeRandom
-Adjusted the jsx to map the dinos to field using levelData.

timjacksonm committed:
refactored the level data based on team feedback.

timjacksonm committed:
updated level data.
-created first 10 levels for beginner difficulty. first 3 levels the winning side is pre determined. After lvl 3 it is random and the legend board is always different from the previous level.

timjacksonm committed:
made more random again

timjacksonm committed:
adjusted naming of props passed to legend.

rankoliang commented on src/components/Legend/index.js:
```diff
@@ -1,29 +1,17 @@
-        <Dino as={TenPointsRedDino} />
-        <span>10</span>
-      </LegendGroup>
+      {legendData.map((set, i) => (
```
It's a bit confusing what `set` means here. You can destructure it to `[dinoId, points]` to make it clearer.

timjacksonm replied:
I seem to forget about doing this a lot. Thanks for the reminder.

rankoliang commented on src/components/Level/index.js:
```diff
@@ -40,38 +41,44 @@ const Level = () => {
-        },
-      };
-    })
+  const [redDinos] = useState(() => {
```
Maybe we could generate the random dinos here instead of on line 97 and with `dinoPerLevelRange`.
It'd be nice if we could just do this instead and have everything in one array.
```js
{redDinos.map(({Component}) => {
    return <Component />
}}
```
Also, whenever the state changes, the dinos will regenerate this way!

rankoliang replied:
Also, we could perhaps shuffle the array so that the same type of dinos aren't always grouped together. Check out this package for an easy way to do it.
https://www.npmjs.com/package/shuffle-array

rankoliang replied:
We could put this into a custom hook so we could reuse it for blueDinos as well if you are up to it.
https://reactjs.org/docs/hooks-custom.html

timjacksonm replied:
Okay so I went back to how you had it originally and tried again. This should be better I hope! I will look into creating a custom hook for how I have it. Just haven't done that part yet. Let me know what you think! Thanks a lot for the guidance.

rankoliang commented on src/components/Level/index.js:
```diff
@@ -40,38 +41,44 @@ const Level = () => {
-        },
-      };
-    })
+  const [blueDinos] = useState(() => {
```
See above

rankoliang commented on src/components/Legend/index.js:
```diff
@@ -1,29 +1,17 @@
+import { dinos, legends } from '../Level/staticLevels'
 
-const Legend = ({ color, reversed }) => {
+const Legend = ({ color, reversed, teamId }) => {
```
I think you can rename `teamId` to `legend` here. (Don't forget to update it in other places)

timjacksonm replied:
Agreed

rankoliang commented on src/components/Level/staticLevels/index.js:
```diff
@@ -0,0 +1,213 @@
+import { ReactComponent as OnePointRedDino } from '../../../icons/dinosaurs/red-one-point.svg';
```
My idea naming these originally was a bit different than yours and I'm not sure if it makes the most sense this way anymore. In each of the SVG files, each dino has an id. Perhaps we can change it by renaming the svgs to `dino-21.svg` and the component to `Dino21`. Here's a list of the names if you want to do that:
- OnePointRedDino: { svg: `dino-30.svg`, component: `Dino30` }
- TwoPointsRedDino: { svg: `dino-20.svg`, component: `Dino20` }
- ThreePointsRedDino: { svg: `dino-24.svg`, component: `Dino24` }
- FivePointsRedDino: { svg: `dino-15.svg`, component: `Dino15` }
- TenPointsRedDino: { svg: `dino-03.svg`, component: `Dino03` }
- OnePointBlueDino: { svg: `dino-09.svg`, component: `Dino09` }
- TwoPointsBlueDino: { svg: `dino-22.svg`, component: `Dino22` }
- ThreePointsBlueDino: { svg: `dino-21.svg`, component: `Dino21` }
- FivePointsBlueDino: I may have messed up and used the same dino for this and three-points-red, so we can get rid of this one or replace it.
- TenPointsBlueDino: { svg: `dino-14.svg`, component: `Dino14` }

timjacksonm replied:
Not quite sure how I could get this to work. Would the dinos variable that we are exporting become what you have above? And no import statement for any svgs?
As for the dino's we have 1, 2, 3, 5, 10 points for each team. That is what I decided we should use.
I am a little confused on what you mean by removing FivePointsBlueDino but leaving the FivePointsRedDino in.
Will need a little more clarification. I am confused. Maybe it could be better! So lets chat more about it.

rankoliang replied:
So instead of 
```js
import { ReactComponent as OnePointRedDino } from '../../../icons/dinosaurs/red-one-point.svg';
```,
we name it like
```js
import { ReactComponent as Dino30 } from '../../../icons/dinosaurs/dino-30.svg';
```
It makes it easier to change the amount of points they are worth if we ever need to make that decision later on. Or use them in other places. Right now, the naming is coupled to points, which it doesn't have to be. Naming it like this will make it more flexible for us. Or if we *do* want to couple the points to the dinos, add the points attribute to each of the dino objects.

We can remove the `blue-five-points' file and when we define the component, reuse the component we imported already, like
```js
const dinos = {
  // previously ThreePointsRedDino
  3: { id: 3, Component: Dino24 },
  // previously FivePointsBlueDIno
  9: { id: 9, Component: Dino24 },
}
```

timjacksonm replied:
Done

rankoliang commented on src/components/Level/staticLevels/index.js:
```diff
@@ -0,0 +1,213 @@
+import { ReactComponent as FivePointsBlueDino } from '../../../icons/dinosaurs/blue-five-points.svg';
+import { ReactComponent as TenPointsBlueDino } from '../../../icons/dinosaurs/blue-ten-points.svg';
+
+export const dinos = {
```
I think this file/folder makes more sense somewhere like `src/data` instead of `src/components/Level/staticLevels`. Also, what do you think about separating out all of the different data into different files? Like `src/data/dinos.js`, `src/data/legends.js` and so on?

timjacksonm replied:
I agree src/data makes more sense as other components will use this data. I can change that.

Potential issue that I am looking for suggestions on. With the current setup legends only displays 3 dinos per legend board. I did this purposely since that is how beginner difficulty will work. I am open to separating the files out but I need some help figuring out how we are going to add in more dinos under legend if the difficulty is Average. And for Hard adding in the skeleton dinos for subtraction. Kind of stumped on what we will do for that.
The links for the level menu I just created will bring you to #/levels/Beginner/:1 for example.

rankoliang replied:
```js
17: {
  7: 2,
  9: 5,
  10: 10,
  11: 20,
},
```
You can just create a new legend with four dinos instead of 3 and reference that. If we want to separate it out by difficulty, we can add nested a nested hash like
```js
const levels = {
  beginner: {
    1: {
      // data
    }
  },
  hard: {
    1: {
      // data
    }
  }
}
```

timjacksonm replied:
done. added both hard and average difficulty. 10 levels each.

rankoliang commented on src/components/Level/staticLevels/index.js:
```diff
@@ -0,0 +1,213 @@
+    '10': { id: 10, Component: TenPointsBlueDino },
+  }
+
+  export const legends = {
```
The formatting is a bit off here (It's tabbed one over). Do you have prettier?

timjacksonm replied:
'9': { id: 9, Component: FivePointsBlueDino },
'10': { id: 10, Component: TenPointsBlueDino },
      ^--- double digits threw it off.
      I do have prettier disabled at the moment. Wasn't sure if the team was using it and didn't want to have to deal with unintentional changes.

rankoliang commented on src/components/Level/staticLevels/index.js:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  export const levels = { //first three levels are fixed winners.
+    ':1': {
```
It's more conventional to have the params as `1` instead of `:1`, so I would change it to that.

timjacksonm replied:
updated.
I also had to change the below code in the level index.js
  const { difficulty } = useParams(); // <--- #/levels/:1
  const levelId = difficulty.split('')[1]; // <--- 1
  const level = levels[levelId]; // <-- levels[1]

rankoliang replied:
How are you visiting the link? If we link users to `/levels/1`, we wouldn't have to process the link like that.

timjacksonm replied:
Oh... yeah I was visiting the wrong link. I was doing #/levels/:1 I figured the semi colon was required since that is how the router is setup. I just changed it and see that it works perfectly fine as levels/1. Will fix.

rankoliang replied:
The `:` just means that the part of the path is dynamic!

rankoliang reviewed (changes requested):
Overall, great job! There is just that bug you need to fix and some minor changes here and there that I would change to make things more clear.

timjacksonm committed:
destructured set to [dinoId, points] to better explain what set is

timjacksonm committed:
changed teamId to legend.

timjacksonm committed:
updated to be 1 intead of :1. also updated index.js for levels to reflect changes

timjacksonm committed:
changed levels[levelId] to levels[difficulty] and removed previous commit changes.

timjacksonm committed:
refactored dino arrays. installed shuffle-array

timjacksonm committed:
changed require to import statement for shuffle.

timjacksonm committed:
changed getRange function to use reduce method.

timjacksonm committed:
changed 5-point-blue.svg to a different dino svg file. Renamed all svgs to correspond with their id in dinos. Re-wrote imports.

timjacksonm committed:
folder structure
-removed staticLevel folder and added data too src/data.
-moved levels, legends to their own js files inside of src/data.
-renamed index.js to dinos.js and left the dinos variable in the file.
-adjusted imports on level index.js and legend index.js

timjacksonm committed:
updated router for difficulty/stage. enabled prettier with double quote jsx. single quote everything else.

timjacksonm committed:
data updates
-added legend boards that include 4 dinos and also 3 dinos & operand. Have it as subtraction for now.
-added dino11-dino15 for dead dinos to be used as operands.
-created 10 levels for average and hard diffuclty. hard lvl 10 is kind of crazy. It is possible to have a negative value I think so probably won't work. We can change this as we get the rest of the game developed.

timjacksonm committed:
data more random

rankoliang commented on src/Router.js:
```diff
@@ -9,7 +9,13 @@ const Router = () => {
           <Home />
         </Route>
-        <Route exact path="/levels/:difficulty">
+        <Route exact path="/levels/beginner/:stage">
```
This should be `<Route exact path"/levels/:difficulty/:stage" />`
You won't have to add a route for each different difficulty this way and you can access it in the component in an easier way.

rankoliang commented on src/components/Level/index.js:
```diff
@@ -11,21 +11,30 @@ import {
-  const { difficulty } = useParams();
-
+  const { stage } = useParams();
+  const difficulty = window.location.href
```
Once you make the changes to the routes above, destructure difficulty from useParams instead.
`const { state, difficulty } = useParams()`

rankoliang reviewed (approved):
One really small change for the routes. Other than that, LGTM! We can squash and merge when you're finished with this.

timjacksonm committed:
updated useParams to include difficulty per suggestion.
